org/mozilla/jss/pkcs11/attrs/CKAttribute.c

```diff
@@ -0,0 +1,271 @@
+    if (field_id == NULL) {
+        goto failure;
+    }
+
+    *ptr = (CK_ULONG)((*env)->GetLongField(env, this, field_id));
+
+    if (JSS_PK11_WrapAttribute(env, this, (void *)ptr, sizeof(*ptr)) == PR_FAILURE) {
+        goto failure;
+    }
+
+    return;
+
+failure:
+    memset(ptr, 0, sizeof(*ptr));
+    free(ptr);
+}
+
+JNIEXPORT void JNICALL
+Java_org_mozilla_jss_pkcs11_attrs_CKAClass_releaseNativeResources(JNIEnv *env, jobject this)
+{
+    CK_ATTRIBUTE_PTR attr = NULL;
+
+    if (JSS_PK11_UnwrapAttribute(env, this, &attr) != PR_SUCCESS || attr == NULL) {
+        return;
+    }
+
+    PR_ASSERT(attr->type == CKA_CLASS);
+    PR_ASSERT(attr->pValue != NULL);
+    PR_ASSERT(attr->ulValueLen == sizeof(CK_ULONG));
+
+    if (attr->pValue != NULL) {
+        memset(attr->pValue, 0, attr->ulValueLen);
+        free(attr->pValue);
+    }
+
+    memset(attr, 0, sizeof(CK_ATTRIBUTE));
+    free(attr);
+
+    return;
+}
+
+/* ===== CKA_KEY_TYPE Attribute ===== */
+
+JNIEXPORT void JNICALL
+Java_org_mozilla_jss_pkcs11_attrs_CKAKeyType_acquireNativeResources(JNIEnv *env, jobject this)
+{
+    jclass this_class;
+    jfieldID field_id;
+    CK_ULONG *ptr = calloc(1, sizeof(CK_ULONG));
+
+    PR_ASSERT(env != NULL && this != NULL && ptr != NULL);
+
+    this_class = (*env)->GetObjectClass(env, this);
+    if (this_class == NULL) {
+        goto failure;
+    }
+
+    field_id = (*env)->GetFieldID(env, this_class, "value", "J");
+    if (field_id == NULL) {
+        goto failure;
+    }
+
+    *ptr = (CK_ULONG)((*env)->GetLongField(env, this, field_id));
+
+    JSS_PK11_WrapAttribute(env, this, (void *)ptr, sizeof(*ptr));
+
+    return;
+failure:
+    memset(ptr, 0, sizeof(*ptr));
+    free(ptr);
+}
+
+JNIEXPORT void JNICALL
+Java_org_mozilla_jss_pkcs11_attrs_CKAKeyType_releaseNativeResources(JNIEnv *env, jobject this)
+{
+    CK_ATTRIBUTE_PTR attr = NULL;
+
+    if (JSS_PK11_UnwrapAttribute(env, this, &attr) != PR_SUCCESS || attr == NULL) {
+        return;
+    }
+
+    PR_ASSERT(attr->type == CKA_KEY_TYPE);
+    PR_ASSERT(attr->pValue != NULL);
+    PR_ASSERT(attr->ulValueLen == sizeof(CK_ULONG));
+
+    if (attr->pValue != NULL) {
+        memset(attr->pValue, 0, attr->ulValueLen);
+        free(attr->pValue);
+    }
+
+    memset(attr, 0, sizeof(CK_ATTRIBUTE));
+    free(attr);
+
+    return;
+}
+
+/* ===== CKA_{ENCRYPT,DECRYPT,...} Usage Attributes ===== */
+
+JNIEXPORT void JNICALL
+Java_org_mozilla_jss_pkcs11_attrs_CKAUsage_acquireNativeResources(JNIEnv *env, jobject this)
+{
+    JSS_PK11_WrapAttribute(env, this, (void *)&JSS_CK_TRUE, sizeof(JSS_CK_TRUE));
+}
+
+JNIEXPORT void JNICALL
+Java_org_mozilla_jss_pkcs11_attrs_CKAUsage_releaseNativeResources(JNIEnv *env, jobject this)
+{
+    CK_ATTRIBUTE_PTR attr = NULL;
+
+    if (JSS_PK11_UnwrapAttribute(env, this, &attr) != PR_SUCCESS || attr == NULL) {
+        return;
+    }
+
+    /* Since this is always a reference to JSS_CK_TRUE, don't free it! */
```

Comment:
@cipherboy: clarify comment. This always _contains_ a reference to `JSS_CK_TRUE`; don't free the reference.